integration/v3_lock_test.go

```diff
@@ -101,6 +101,83 @@ func TestMutexSessionRelock(t *testing.T) {
 	}
 }
 
+// TestMutexLockOnlyAfterOldestReleasesLock ensures a series of waiters
+// obtain lock only after the oldest waiter releases lock.
+// e.g a series of Waiters A <- B <- C where A holds the lock
+// and B waits before C. The failure of B won't trigger C to obtain the lock.
+func TestMutexLockOnlyAfterOldestReleasesLock(t *testing.T) {
+	clus := NewClusterV3(t, &ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+
+	cctx, ccancel := context.WithCancel(context.Background())
+
+	cli := clus.Client(0)
+	s1, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	s2, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+	s3, err := concurrency.NewSession(cli)
+	if err != nil {
+		t.Error(err)
+	}
+
+	m := concurrency.NewMutex(s1, "test-mutex")
+	if err := m.Lock(cctx); err != nil {
+		t.Fatal(err)
+	}
+
+	m2 := concurrency.NewMutex(s2, "test-mutex")
+	m2donec := make(chan struct{})
+	go func() {
+		defer close(m2donec)
+		m2.Lock(cctx)
+	}()
+
+	// ensure m2 waits before m3
+	time.Sleep(100 * time.Millisecond)
+	m3 := concurrency.NewMutex(s3, "test-mutex")
+	m3donec := make(chan struct{})
+	go func() {
+		defer close(m3donec)
+		m3.Lock(cctx)
+	}()
+
+	// close s2 to simulate a node failure
+	// which results deletion of m2 waiter due lease being expired.
+	s2.Close()
```

Comment:
this should wait until there's another event on the prefix so that `m3` has definitely tried to acquire the lock; otherwise this can vanish before the lock acquire happens